### ISSUE TITLE
opencoarray and MPI_Init, MPI_Finalize

### ISSUE THREAD

**Issue by mexas** [sourceryinstitute/OpenCoarrays#349]
This is about coarray programs which use MPI calls directly, e.g. MPI/IO calls.

It seems with Cray and Intel such programs must have explicit MPI_Init and
MPI_Finalize. Not so with OpenCoarrays. I can check with MPI_Initialized and
call MPI_Init only if MPI hasn't been initialised already. That works ok.

But a check for MPI_Finalized is useless, because it will always return false
until the program terminates and OpenCoarrays call MPI_Finalize behind
the scene, right?

Anyway, I'm just wondering which way is the "correct" way - Cray/Intel or OpenCoarrays?

From the Cray man page on coarrays:
*quote*
   Mixing PGAS with MPI
       UPC, Coarray C++, and Fortran are compatible with MPI. The programmer is responsible for
       calling MPI_Init(3) and MPI_Finalize(3) from their main program. All of the PGAS languages
       automatically initialize themselves prior to execution of the main program and finalize
       themselves upon completion of the main program, so unlike MPI there is no need to
       explicitly call an initialization or finalization function for PGAS.
*end quote*

so they want the use to explicitly add MPI_Init/MPI_Finalize.

Intel do not suport MPI/coarrays. So although it works, there is not official Intel
advice on how this should be done.

Thanks

Anton

**Comment by afanfa** [comment]
Hi Anton, you made a really good point.
Opencoarrays do invoke MPI_Finalize at the end of the execution and for mixed MPI/CAF codes I usually leave Opencoarrays to do the initialization.
Opencoarrays checks whether MPI has been already initialized but apparently it is not working properly, probably because the compiler invokes caf_init before the main program.

**Comment by vehre** [comment]
Opencoarrays checks whether MPI was already initialized during its caf_init(), correct.
When gfortran compiles the main-program there is no way around this caf_init() and it taking the responsibility for initializing mpi. The check in caf_init(), whether mpi is initialized already, is there to be able to use coarray enabled code from programs, whose main-program is not fortran or which was not compiled with the coarray-option. 
So this is mainly for libraries, that use mpi-coarray where the non-library code is using mpi already. The library then is responsible for starting the caf-runtime and finalizing it. To my knowledge is this not very well tested. Furthermore are you stepping into non-standard area here, because caf_init and caf_finalize are non-standard routines and may change without notice.

Hope this helps,
Andre

**Comment by zbeekman** [comment]
HI Anton/@mexas, I agree with Alessandro that this issue deserves some more attention.

An additional note: In the future could you please make some attempt/semblance of filling out the relevant portions of the Github issue template that we've added? I know it needs some cleaning up and tightening up, but, for example, labeling your issue according to the instructions of the template would have made it immediately clear that this was a question about usage, rather than a request for enhancement (RFE) or problem report. As I said, the template needs some additional attention, and we would welcome any feedback, suggestions and pull requests to improve it. (It can be found in the .github directory in the projects root directory: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/.github/ISSUE_TEMPLATE.md) In this particular case, it doesn't matter that much, but in other cases, ensuring that you include the versions of OpenCoarrays, the fortran compiler, the MPI library/runtime, and OS etc. is extremely helpful for us, and prevents us from wasting our time and yours by having to ask for this information.

**TL;DR: Please use the issue template and if you are unhappy with it, provide some feedback to help us improve it.**

Many thanks!

**Comment by vehre** [comment]
Hi Anton,

well, rereading you question, I figured, I missed the target with my answer.
Yes, you are right, that MPI_initialized() will always return false when the main-program was compiled using -fcoarray=lib as well as MPI_Finalized() will always return false.

Those checks only make sense, when you use coarrays only in certain portions of your program and do not compile everything with -fcoarray=lib. You are then responsible for managing the caf-runtime yourself, which I do not recommend, if it is not absolutely necessary, and there my former answer might be of help, if you choose to take the bother.

- Andre

**Comment by mexas** [comment]
RE template - it's far too long and cumbersome. I'm unlikely to do it unless it's reduced to 1-2 lines. I have to work with a great number of OS projects, i.e. submitting problem reports, and your template is by far the longest.

**Comment by zbeekman** [comment]
@mexas I'm closing this issue since it appears to be resolved. If you feel that your question has not been adequately addressed please  feel free to reopen it.
